src/valkey-mcp-server/awslabs/valkey_mcp_server/tools/semantic_search.py

```diff
@@ -0,0 +1,627 @@
+from valkey.exceptions import ValkeyError
+
+
+# Initialize the embeddings provider based on environment configuration
+_embeddings_provider = create_embeddings_provider()
+
+
+def _get_collection_index_name(collection: str) -> str:
+    """Get the Valkey index name for a collection."""
+    return f'semantic_collection_{collection}'
+
+
+def _get_document_key(collection: str, document_id: str) -> str:
+    """Get the Valkey key for a document."""
+    return f'semantic_collection_{collection}:doc:{document_id}'
+
+
+def _index_exists(
+    conn: Union[Valkey, ValkeyCluster],
+    collection: str,
+) -> bool:
+    """Check if a collection index exists."""
+    index_name = _get_collection_index_name(collection)
+    try:
+        conn.execute_command('FT.INFO', index_name)
+        return True
+    except ValkeyError:
+        return False
+
+
+@mcp.tool()
+async def add_documents(
+    collection: str,
+    documents: List[Dict[str, Any]],
+    text_fields: Optional[List[str]] = None,
+    embedding_dimensions: Optional[int] = None,
+) -> Dict[str, Any]:
+    """Add documents to a collection with automatic embedding generation.
+
+    This tool stores documents in a searchable collection, automatically generating
+    vector embeddings from the specified text fields. Documents can contain any fields,
+    but must have an 'id' field.
+
+    The embedding provider is configured via the EMBEDDINGS_PROVIDER environment variable.
+    The embedding dimensions are automatically detected from the provider's output.
+    If embedding_dimensions is not specified, it will be determined from the first
+    embedding generated.
+
+    Args:
+        collection: Name of the collection (e.g., "research_papers", "customer_reviews")
+        documents: List of documents, each must have an 'id' field
+        text_fields: Fields to use for embedding generation (default: ["content"])
+        embedding_dimensions: Vector dimensions (auto-detected if not specified)
+
+    Returns:
+        Summary of the operation including number of documents added and provider info, or in the case
+        of an error, the error message, in a similar structure.  In both cases, the "status" and "added" fields
+        indicate the overall result of the operation and the number of documents successfully added.
+
+    Example:
+        result = await add_documents(
+            collection="research_papers",
+            documents=[
+                {
+                    "id": "paper_1",
+                    "title": "AI in Healthcare",
+                    "content": "Machine learning is transforming medical diagnosis...",
+                    "author": "Dr. Smith",
+                    "year": 2024
+                }
+            ],
+            text_fields=["title", "content"]
+        )
+    """
+    if text_fields is None:
+        text_fields = ['content']
+
+    try:
+        r = ValkeyConnectionManager.get_connection(decode_responses=True)
```

Comment:
```suggestion
        v = ValkeyConnectionManager.get_connection(decode_responses=True)
```
i assume r is for redis. We should try to be valkey first as much as possible.